api/datasets.go

```diff
@@ -497,6 +410,141 @@ func (h *DatasetHandlers) initHandler(w http.ResponseWriter, r *http.Request) {
+			data, err = ioutil.ReadAll(datafile)
+			if err != nil {
+				err = fmt.Errorf("error reading dataset file: %s", err)
+				return
+			}
+			if err = dsutil.UnmarshalYAMLDatasetPod(data, dsp); err != nil {
+				err = fmt.Errorf("error unmarshaling yaml file: %s", err)
+				return
+			}
+		case ".json":
+			if err = json.NewDecoder(datafile).Decode(dsp); err != nil {
+				err = fmt.Errorf("error decoding json file: %s", err)
+				return
+			}
+		}
+	}
+
+	tfFile, _, err := r.FormFile("transform")
+	if err != nil && err != http.ErrMissingFile {
+		err = fmt.Errorf("error opening transform file: %s", err)
+		return
+	}
+	if tfFile != nil {
+		// TODO - this assumes a skylark / starlark transform file
+		f, e := ioutil.TempFile("", "transform")
+		if e != nil {
+			err = e
+			return
+		}
+		rmFiles = append(rmFiles, f)
+		io.Copy(f, tfFile)
+		if dsp.Transform == nil {
+			dsp.Transform = &dataset.TransformPod{}
+		}
+		dsp.Transform.Syntax = "skylark"
+		dsp.Transform.ScriptPath = f.Name()
+	}
+
+	vizFile, _, err := r.FormFile("viz")
+	if err != nil && err != http.ErrMissingFile {
+		err = fmt.Errorf("error opening viz file: %s", err)
+		return
+	}
+	if vizFile != nil {
+		// TODO - this assumes an html viz file
+		f, e := ioutil.TempFile("", "viz")
+		if e != nil {
+			err = e
+			return
+		}
+		rmFiles = append(rmFiles, f)
+		io.Copy(f, vizFile)
+		if dsp.Viz == nil {
+			dsp.Viz = &dataset.Viz{}
+		}
+		dsp.Viz.Format = "html"
+		dsp.Viz.ScriptPath = f.Name()
+	}
+
+	dsp.Peername = r.FormValue("peername")
+	dsp.Name = r.FormValue("name")
+	dsp.BodyPath = r.FormValue("body_path")
+
+	bodyfile, bodyHeader, err := r.FormFile("body")
+	if err != nil && err != http.ErrMissingFile {
+		err = fmt.Errorf("error opening body file: %s", err)
+		return
+	}
+	if bodyfile != nil {
+		path := filepath.Join(os.TempDir(), bodyHeader.Filename)
+		f, e := os.Create(path)
+		if e != nil {
+			err = fmt.Errorf("error writing body file: %s", e.Error())
+			return
+		}
+		rmFiles = append(rmFiles, f)
+		io.Copy(f, bodyfile)
+		f.Close()
+		dsp.BodyPath = path
+	}
+
+	return
+}
+
+func addBodyFile(res *repo.DatasetRef) error {
+	if res.Dataset.Structure.Format == dataset.JSONDataFormat.String() {
+		// TODO - this'll only work for JSON responses
+		data, err := ioutil.ReadAll(res.Dataset.Body.(io.Reader))
+		if err != nil {
+			return err
+		}
+		res.Dataset.Body = json.RawMessage(data)
+	} else {
+		if file, ok := res.Dataset.Body.(cafs.File); ok {
+			in := &dataset.Structure{}
+			if err := in.Decode(res.Dataset.Structure); err != nil {
+				return err
+			}
+
+			st := &dataset.Structure{}
+			st.Assign(in, &dataset.Structure{
+				Format: dataset.JSONDataFormat,
+				Schema: in.Schema,
+			})
+
+			data, err := actions.ConvertBodyFile(file, in, st, 0, 0, true)
+			if err != nil {
+				return fmt.Errorf("converting body file to JSON: %s", err)
+			}
+			res.Dataset.Body = json.RawMessage(data)
+		} else {
+			log.Error("response body isn't a cafs.File")
```

Comment:
This error handling is too far from its cause. Move it up to where the conversion happens. Put the success case underneath it.